src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs

```diff
@@ -235,40 +233,231 @@ internal void WriteTopLevelProperty(ODataProperty property)
             IDuplicatePropertyNameChecker duplicatePropertyNameChecker,
             ODataResourceMetadataBuilder metadataBuilder)
         {
-            WriterValidationUtils.ValidatePropertyNotNull(propertyInfo);
+            ValidatePropertyInfo(propertyInfo, owningType, isTopLevel, duplicatePropertyNameChecker);
 
-            string propertyName = propertyInfo.Name;
+            if (currentPropertyInfo.MetadataType.IsUndeclaredProperty)
+            {
+                WriteODataTypeAnnotation(propertyInfo, isTopLevel);
+            }
 
-            if (this.JsonLightOutputContext.MessageWriterSettings.Validations != ValidationKinds.None)
+            WriteInstanceAnnotation(propertyInfo, isTopLevel, currentPropertyInfo.MetadataType.IsUndeclaredProperty);
+
+            ODataStreamPropertyInfo streamInfo = propertyInfo as ODataStreamPropertyInfo;
+            if (streamInfo != null && !(this.JsonLightOutputContext.MetadataLevel is JsonNoMetadataLevel))
             {
-                WriterValidationUtils.ValidatePropertyName(propertyName);
+                Debug.Assert(!isTopLevel, "Stream properties are not allowed at the top level.");
+                WriteStreamValue(streamInfo, propertyInfo.Name, metadataBuilder);
             }
+        }
 
-            if (!this.JsonLightOutputContext.PropertyCacheHandler.InResourceSetScope())
+        /// <summary>
+        /// Asynchronously writes an <see cref="ODataProperty" /> to the given stream. This method creates an
+        /// async buffered stream and writes the property to it.
+        /// </summary>
+        /// <param name="property">The property to write.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal Task WriteTopLevelPropertyAsync(ODataProperty property)

```

Comment:
>Task [](start = 17, length = 4)

why no async key word?

Reply:
@xuzhg Eliding is recommended when a method is a pass-through. See this article on the subject
https://blog.stephencleary.com/2016/12/eliding-async-await.html#recommended-guidelines:~:text=Recommended%20Guidelines
This method qualifies as a pass-through since it merely calls another method to do the job.
Eliding will save us the overhead of creating and later garbage-collecting the state machine.
For that reason, I'm eliding the `async` and `await` whenever the method is just a pass-through